lib/pact/api/exec_command_request.ex

```diff
@@ -0,0 +1,315 @@
+  end
+
+  def add_keypairs(%__MODULE__{}, _keypairs), do: {:error, [keypairs: :not_a_list]}
+  def add_keypairs({:error, reason}, _keypairs), do: {:error, reason}
+
+  @impl true
+  def add_signer(%__MODULE__{signers: nil} = cmd_request, %Signer{} = signer),
+    do: %{cmd_request | signers: SignersList.new([signer])}
+
+  def add_signer(%__MODULE__{signers: signer_list} = cmd_request, %Signer{} = signer) do
+    %SignersList{signers: signers} = signer_list
+    %{cmd_request | signers: SignersList.new(signers ++ [signer])}
+  end
+
+  def add_signer(%__MODULE__{}, _signer), do: {:error, [signer: :invalid]}
+  def add_signer({:error, reason}, _signer), do: {:error, reason}
+
+  @impl true
+  def add_signers(%__MODULE__{} = cmd_request, %SignersList{} = list),
+    do: %{cmd_request | signers: list}
+
+  def add_signers(%__MODULE__{} = cmd_request, signers) do
+    case SignersList.new(signers) do
+      %SignersList{} = signers -> %{cmd_request | signers: signers}
+      {:error, reason} -> {:error, [signers: :invalid] ++ reason}
+    end
+  end
+
+  def add_signers({:error, reason}, _signers), do: {:error, reason}
+
+  @impl true
+  def build(
+        %__MODULE__{
+          network_id: %NetworkID{} = network_id,
+          meta_data: %MetaData{chain_id: chain_id},
+          keypairs: keypairs,
+          code: code,
+          data: data
+        } = cmd_request
+      ) do
+    with {:ok, payload} <- create_payload(code, data),
+         {:ok, cmd} <- command_to_json_string(payload, cmd_request),
+         {:ok, sig_commands} <- sign_commands([], cmd, keypairs),
+         {:ok, hash} <- get_unique_hash(sig_commands),
+         {:ok, signatures} <- get_signatures(sig_commands, []),
+         {:ok, command} <- create_command(hash, signatures, cmd) do
+      %CommandRequest{cmd: command, network_id: network_id, chain_id: chain_id}
+    end
+  end
+
+  def build(_module), do: {:error, [exec_command_request: :invalid_payload]}
+
+  defp validate_keypair(%KeyPair{
+         pub_key: pub_key_arg,
+         secret_key: secret_key_arg,
+         clist: clist_arg
+       }) do
+    with {:ok, pub_key} <- validate_key({:pub_key, pub_key_arg}),
+         {:ok, secret_key} <- validate_key({:secret_key, secret_key_arg}),
+         {:ok, clist} <- validate_optional_caps_list({:clist, clist_arg}) do
+      %KeyPair{pub_key: pub_key, secret_key: secret_key, clist: clist}
+    end
+  end
+
+  @spec validate_key(arg :: arg()) :: arg_validation()
+  defp validate_key({_arg, key}) when is_binary(key) and byte_size(key) == 64, do: {:ok, key}
+  defp validate_key({arg, _key}), do: {:error, [{arg, :invalid}]}
+
+  @spec validate_optional_caps_list(arg :: arg()) :: arg_validation()
+  defp validate_optional_caps_list({arg, clist}) do
+    case clist do
+      %OptionalCapsList{} = optional_caps -> {:ok, optional_caps}
+      %CapsList{} = caps_list -> {:ok, OptionalCapsList.new(caps_list)}
+      _error -> {:error, [{arg, :invalid}]}
+    end
+  end
```

Comment:
@EdwinGuayacan @Odraxs IMO, this is not a place to perform these validations. Actually, this is already validated when initializing a `KeyPair`

Reply:
It was already validated that a KeyPair was received, however, it was possible that a Keypair of these types was sent to the function add_keypair():
* KeyPair{ }
* KeyPair{pub_key: "invalid pub_key invalid", secret_key: "invalid", clist: "invalid"}

Reply:
@juanhurtado10 We implement the validation because if we send a Keypair structure instead of one created by `KeyPair.new()`, line `#294 {:ok, sign_command} = Sign.sign(cmd, keypair)` could fail with an unrecoverable erlang error.
Is there another way to handle this ?

Reply:
@EdwinGuayacan @Odraxs Let me explain it a little bit.

![image](https://user-images.githubusercontent.com/1649973/204016914-f4edd3d5-6039-434d-9011-c518d92e80cb.png)

Receiving a `%KeyPair{}` structure as a function parameter already means that the attributes (pub_key, secret_key, clist) were validated when invoking `KeyPair.new/1`. Why is it necessary to validate it again?